Move docs site markdown files into main repo (#162)
tigerbeetle/tigerbeetle: docs/reference/transfers.md

# Transfers

> 

Reviewer: We need a general overview of what a TigerBeetle transfer is. Something along the lines of:

```suggestion
A `Transfer` is an immutable record of a financial transaction between two accounts.

`Transfer`s use double-entry accounting: each transfer updates the balance of the debit account and the credit account.
```

TigerBeetle uses the same data structures internally and
externally. This means that sometimes you need to set temporary values
for fields that TigerBeetle, not you (the user), are responsible.

### Updates

Transfer fields *cannot be changed by the user* after
creation. However, debits and credits fields are updated by

> TigerBeetle as transfers move money to and from an account.

Reviewer: I added "immutable" to the introduction section.

```suggestion
```

Author: I'm not sure what is changed here now

## Fields

> 

Reviewer: (See comments for corresponding fields in `accounts.md`).

Author: :+1:

### `id`

This is the unique identifier, the primary key, of the transaction.

As an example, you might generate a UUID (encoded as an integer) to
identify each transaction.

Constraints:

* Type is 128-bit unsigned integer (16 bytes)
* Must not be zero or `2^128 - 1`
* Must be unique

### `debit_account_id`

> The account to pull an amount from.

Reviewer: I think the pull/send terminology is misleading — accounts' posted amounts only ever increase.

Author: Removed :+1:

Constraints:

* Type is 128-bit unsigned integer (16 bytes)
* Must match an existing account
* Must not be the same as `credit_account_id`

### `credit_account_id`

The account to send an amount to.

Constraints:

* Type is 128-bit unsigned integer (16 bytes)
* Must match an existing account
* Must not be the same as `debit_account_id`

### `user_data`

This is an optional second identifier to link this transfer to an
external entity.

As an example, you might use a UUID (encoded as an integer) that
ties together a group of transfers.

Constraints:

* Type is 128-bit unsigned integer (16 bytes)
* May be zero

### `reserved`

Reserved for future use.

Constraints:

* Type is 128-bit unsigned integer (16 bytes)
* Must be zero

### `pending_id`

If this transfer will post or void a pending transfer, this field is
for the id of that pending transfer. (You specify whether it will
post or void a pending transfer in the [`flags`](#flags) field

> documented below).

Reviewer: ```suggestion
If this transfer will post or void a pending transfer, `pending_id` references that pending transfer.
If this is not a post or void transfer, it must be zero.

See also:
- [`flags.post_pending_transfer`](#flags.post_pending_transfer)
- [`flags.void_pending_transfer`](#flags.void_pending_transfer)
```

Author: Taken.

> 

Reviewer: For explaining pending/posted/voided transfers, what do you think of having worked-out examples like this (hide in `<details>` even in the actual document, so they don't take up too much space):

<details><summary>Example pending transfer with partial posting</summary>

Consider a series of transfers:

| Transfer | `amount` | `flags`                       | `debit_account_id` | `credit_account_id` | `pending_id`  |
| ---:     | ---:     | :---                          | ---:               | ---:                | ---:          |
| `T₁`     | 30      | `flags.pending`               | `A`                | `B`                 |              |
| `T₂`     | 20        | `flags.post_pending_transfer` | `A`                | `B`                 | `T₁`          |

The transfers effects on the accounts are shown:

| Time | Account | `debits_pending` | `debits_posted` | `credits_pending` | `credits_posted` |
| ---:              | ---: | ---:             | ---:            | ---:              | ---:             |
| before `T₁`| `A` | 0                | 0               | 0                 | 0                |
| before `T₁`| `B` | 0                | 0               | 0                 | 0                |
| after `T₁`              | `A` | 30              | 0               | 0                 | 0                |
| after `T₁`              | `B` | 0                | 0               | 30               | 0                |
| after `T₂`              | `A` | 0              | 20               | 0                 | 0                |
| after `T₂`              | `B` | 0                | 0               | 0               | 20                |

</details>

Reviewer 2: This is a great idea. Would be good to put the amount always on the right column.

The shorthand for journal entries (e.g. when writing accounting exams, or when lecturers write their notes) is:

```
DR    Bank                    100
  CR    Loan from Co-Founder      100
```

Reviewer 2: i.e. “Account before amount”.

Author: Good rule! :D And yes DJ I'll add that!

Constraints:

* Type is 128-bit unsigned integer (16 bytes)
* Must be zero if neither void nor pending transfer flag is set
* Must match an existing transfer if non-zero

### `timeout`

This field relates to the [`pending` flag](#pending-flag) described

> below. It is a duration expressed in nanoseconds.

Reviewer: ```suggestion
`timeout` is the interval (in nanoseconds) after a [`pending`](#flags.pending) transfer's creation that it may be posted or voided.
If the timeout expires and the pending transfer has not already been posted or voided, the pending transfer is voided automatically.

Non-pending transfers cannot have a timeout.
```

Constraints:

* Type is 64-bit integer (8 bytes)
* Must be zero if `pending` flag is *not* set
* Must be non-zero if `pending` flag *is* set

### `ledger`

Identifier used to enforce transfers between the same ledger.

Constraints:

* Type is 32-bit unsigned integer (4 bytes)
* Must not be zero
* Must match the `ledger` value on the accounts referenced in `debit_account_id` **and** `credit_account_id`

### `code`

A user-defined enum representing the reason for (or type of) the
transfer.

Constraints:

* Type is 16-bit unsigned integer (2 bytes)
* Must not be zero

### `flags`

> Specifies behavior during transfers.

Reviewer: ```suggestion
Specifies transfer behavior.
```

Author: Taken

Constraints:

* Type is 16-bit unsigned integer (2 bytes)

> #### `linked` flag

Reviewer: ```suggestion
#### `flags.linked`
```

and likewise for other flags.

Author: If this is a Zig thing this may not be very clear to non-zig users. It is less clear to me than "`linked` flag"

Author: nvm

When the linked flag is specified, it links an transfer with the next
transfer in the batch, to create a chain of transfer, of arbitrary
length, which all succeed or fail in creation together. The tail of a
chain is denoted by the first transfer without this flag. The last
transfer in a batch may therefore never have the `linked` flag set as
this would leave a chain open-ended.

Multiple chains or individual transfers may coexist within a batch to
succeed or fail independently. Transfers within a chain are executed
within order, or are rolled back on error, so that the effect of each
transfer in the chain is visible to the next, and so that the chain is
either visible or invisible as a unit to subsequent transfers after the
chain. The transfer that was the first to break the chain will have a
unique error result. Other transfers in the chain will have their error

> result set to `linked_event_failed`.

Reviewer: ```suggestion
When the linked flag is specified, it links a transfer with the next
transfer in the batch, to create a chain of transfer, of arbitrary
length, which all succeed or fail in creation together. The tail of a
chain is denoted by the first transfer without this flag. The last
transfer in a batch may therefore never have the `linked` flag set as
this would leave a chain open-ended.

Multiple chains or individual transfers may coexist within a batch to
succeed or fail independently. Transfers within a chain are executed
within order, or are rolled back on error, so that the effect of each
transfer in the chain is visible to the next, and so that the chain is
either visible or invisible as a unit to subsequent transfers after the
chain. The transfer that was the first to break the chain will have a
unique error result. Other transfers in the chain will have their error
result set to `linked_event_failed`.
```

This explanation (and the one in `accounts.md`) would be well-served by a diagram or perhaps a table to explain the concept visually, e.g:

---

Consider creating this batch of transfers:

| index within batch | transfer | `flags.linked` |
| ---: | ---: | --- |
| `0` | `A` | `false` |
| `1` | `B` | `true` |
| `2` | `C` | `true` |
| `3` | `D` | `false` |
| `4`| `E` | `false` |

If any of transfers `B`, `C`, or `D` fail (for example, due to `error.exceeds_credits`), then `B`, `C`, and `D` will all fail.

Transfers `A` and `E` fail or succeed independently of `B`, `C`, and `D`.

Author: Stolen, thanks!

#### `pending` flag

This flag  commit for this transfer. The transfer is
not complete until a transaction with the same id is sent with the
`post_pending_transfer` or `void_pending_transfer` flag set (both
described below).

> When this flag is on, the [`timeout`](#timeout) field must be non-zero.

Reviewer: ```suggestion
A `pending` transfer reserves its `amount` in the debit/credit accounts' `debits_pending`/`credits_pending` respectively.

The pending transfer is complete when the first of the following events occur:
- If a corresponding [`post_pending_transfer`](#flags.post_pending_transfer) is committed, some or all of the pending transfer's reserved funds are posted, and the remainder (if any) is restored to the original accounts.
- If a corresponding [`void_pending_transfer`](#flags.void_pending_transfer) is committed, the pending transfer's reserved funds are restored to their original accounts.
- If the pending transfer's timeout expires, the pending transfer's reserved funds are restored to their original accounts.

When this flag is set, the [`timeout`](#timeout) field must be non-zero.
```

Author: Taken

#### `post_pending_transfer` flag

This flag causes the transfer to move from pending to posted. The

> transfer succeeds.

Reviewer: ```suggestion
A `post_pending_transfer` causes the referenced pending transfer to "post", transferring some or all of the money the pending transfer reserved to its destination, and restoring the remainder (if any) to its origin accounts.

The `amount` of a `post_pending_transfer` must be less-than or equal-to the pending transfer's amount.
- If the posted `amount` is 0, the full pending transfer's amount is posted.
- If the posted `amount` is nonzero, then the posted transfer's amount is posted (reword this please!), and the remainder is restored to its original accounts.
```

Author: done, thanks!

> 

Reviewer: Note constraints that when `post_pending_transfer` is set:
- `pending_id` must be a pending transfer.
- `debit_account_id` can either be zero, or it can match the pending transfer's `debit_account_id`.
- `credit_account_id` can either be zero, or it can match the pending transfer's `credit_account_id`.
- `ledger` can either be zero, or it can match the pending transfer's `ledger`.
- `code` can either be zero, or it can match the pending transfer's `code`.
- `post_pending_transfer` is mutually exclusive with `void_pending_transfer`.

(likewise applicable to `void_pending_transfer`)

Author: Done, thanks!

#### `void_pending_transfer` flag

This flag causes the transfer to move from pending to nothing. The
transfer does not go through.

### `amount`

How much should be sent from the `debit_account_id` account to the
`credit_account_id` account.

* Type is 64-bit unsigned integer (8 bytes)

> * Must not be zero

Reviewer: `amount` can be zero for posted transfers, in which case the full pending amount is posted.

### `timestamp`

Time the account was created. This is set by TigerBeetle. The format
is UNIX timestamp in nanoseconds.

It is set by TigerBeetle the moment the account is committed (created
by consensus).

Constraints:

* Type is 64-bit unsigned integer (8 bytes)
* User sets to zero on creation

> ## Internals

Reviewer: Also link to the state machine. Reference functions `create_transfer()` and `post_or_void_pending_transfer()`.

  committed, some or all of the pending transfer's reserved funds are
  posted, and the remainder (if any) is restored to the original
  accounts.
* If a corresponding
  [`void_pending_transfer`](#flags.void_pending_transfer) is
  committed, the pending transfer's reserved funds are restored to
  their original accounts.
* If the pending transfer's timeout expires, the pending transfer's
  reserved funds are restored to their original accounts. (When `flags.pending`
  is set, the [`timeout`](#timeout) field must be non-zero.)

#### `flags.post_pending_transfer`

This flag causes the pending transfer (referred to by
[`pending_id`](#pending_id)) to "post", transferring some or all of the money the
pending transfer reserved to its destination, and restoring the
remainder (if any) to its origin accounts. The `amount` of a
`post_pending_transfer` must be less-than or equal-to the pending
transfer's amount.

* If the posted `amount` is 0, the full pending transfer's amount is
  posted.
* If the posted `amount` is nonzero, then only this amount is posted,
  and the remainder is restored to its original accounts. It must be
  less than or equal to the pending transfer's amount.

> Additionally, when this field is set:

Reviewer: ```suggestion
Additionally, when `flags.post_pending_transfer` is set:
```
Be explicit, since we have referred to other fields.

* `pending_id` must reference a pending transfer.
* `flags.void_pending_transfer` must not be set.

And the following fields may either be zero, otherwise must match the
value of the pending transfer's field:

* `debit_account_id`
* `credit_account_id`
* `ledger`
* `code`

#### `flags.void_pending_transfer`

This flag causes the pending transfer (referred to by
[`pending_id`](#pending_id)) to void. The pending transfer's
referenced debit/credit accounts'
[`debits_pending`](./accounts.md#debits_pending)/[`credits_pending`](./accounts.md#credits_pending)
fields .

Additionally, when this field is set:

* `pending_id` must reference a pending transfer.
* `flags.post_pending_transfer` must not be set.

And the following fields may either be zero, otherwise must match the
value of the pending transfer's field:

* `debit_account_id`
* `credit_account_id`
* `ledger`

> * `code`

Reviewer: ```suggestion
* `code`
* `amount`
```

Author: Huh? Didn't we just say this was allowed to be less than or equal to the original amount? Not that it must be zero or the original amount.

Reviewer: The "less-than or equal to the original amount" is only for `flags.post_pending_transfer`. `flags.void_pending_transfer` always voids the entire amount.

Reviewer: See also: https://github.com/tigerbeetledb/tigerbeetle/blob/ebad9ef5c85d0debc9fb34469f78e3c4e55dd7cf/src/state_machine.zig#L738-L743

Author: Oooops sorry I misread which section this is. Good point!

### `amount`

This is how much should be debited from the `debit_account_id` account
and credited to the `credit_account_id` account.

* Type is 64-bit unsigned integer (8 bytes)
* Must not be zero, unless marking a pending transfer as posted

### `timestamp`

This is the time the account was created, as nanoseconds since

> UNIX epoch.

Reviewer 3: The time of the "transfer".
I'd suggest rephrasing it with the same expression from `accounts.md`.

Author: Fixed, thank you!
